### ISSUE TITLE
Proposal: remote vocabularies via urls

### ISSUE THREAD

**Issue by lolbinarycat** [factor/factor#2340]
Currently, all the vocabularies you might want to use have to be on your drive. Most other languages have some sort of package system. I particularly like Go's system of not relying on a central database. I think we could go one step further and not rely on a specific **protocol**.

Many of the details are still up in the air, I just want to get some input on this before I commit to anything.

Here's a **rough** implementation:
```factor
USING: urls curl combinators kernel arrays accessors
sequences io.files io.directories io.pathnames parser
lexer vocabs.loader ;

IN: vocabs.loader.remote

: trim-slashes ( seq -- seq )
    [ CHAR: / = ] trim ;

: cache-path-for-url ( url -- string )
    [ protocol>> ] [ host>> ] [ path>> trim-slashes ] tri
    3array "resource:cache" prefix "/" join ; 

: url-cached? ( url -- ? )
    cache-path-for-url exists? ;

: download-file-to ( url path -- )
    make-parent-directories absolute-path curl-download-to ;

: get-file-from-url ( url -- path )
    dup cache-path-for-url dup exists?
    [ 2dup download-file-to ] unless nip ;

: run-url ( url -- )
    get-file-from-url run-file ;
    
SYNTAX: REMOTE: parse-token >url get-file-from-url path-components add-vocab-root ;
```

**Comment by mrjbq7** [comment]
Pretty cool!

We might want to use ``cache-directory`` since ``resource:`` at some point for system-wide installations won't be necessarily writable.  That has the benefit of surviving restarts, unlike ``temp-directory``, but also might be deleted so... perhaps in that case something like ``~/.factor/vocab-roots`` is a better way...

How do the other languages handle versions, or updating the code represented by these URLs?  Do they download once and then consider that the version unless explicitly updated?  Or have the version embedded in the URL?

**Comment by lolbinarycat** [comment]
Well, it is a cache, so it would make sense that it can be deleted. as for versions, that's a question factor as a whole needs to answer. In working on this I looked through some of the internals for vocab loading, and it isn't all that neat (stuff like .private being hardcoded). Makes me consider broader package reform.

**Comment by timor** [comment]
In my opinion that is a very difficult problem.  One which I have not seen a single language/implementation find an optimum solution for regarding at least the following issues:
1. How are dependencies specified if they involve different versions?
2. How are physical locations related to these "logical" dependencies?
3. How does the programmer of an application specify which are the correct things to have?
4. How does an OS distribution which has its own package handling relate to these specifications, etc...

I personally strongly prefer approaches which separate in-language lookup from on-filesystem/in-internet locations as much as possible.  This is admittedly coming from a NixOS perspective, which takes this to the extreme, but in general I find it much easier to deal with the "configuration" of how the implementation finds it's stuff (vocab roots, lookup paths, priorities, etc.) on the "outside" of the implementation.  E.g. currently factor has the "work:".. prefixes, but they are hard-coded locations, and all the others may be read-only locations, depending on how Factor is installed. (Under NixOS, factor has to be patched to let "work:" point to somewhere else because anything is installed in a read-only location by default).

In my experience, the approach that Emacs takes, or for example how Git is configured, is the easiest to work with (has its own shortcomings, of course).  There is a clear interface which specifies where on the system to look for things in which order (distribution-pre-installed, site-wide, user-wide, project-wide, ...), and that usually allows "layering" the correct environment together quite nicely.  The biggest problems I have had were usually due to some package wanting to configure everything itself, the worst case being that it only works when a dependency has to be 1. at a place in the system determined by the package and 2. at a time during installation chosen by a package.

My suggestions would look like this:
1. Have the existing "work:", "resource:", etc prefixes be "virtual" locations which are composed of different standard system locations
2. Allow adding configurable, possibly remote urls to that composition, making sure to keep control over priorities
3. If needed, have a dependency system where a package can "request" certain versions of a vocab (is there supposed to be a global package namespace though, like in the java world?), but let the "outside" be responsible for resolving those, maybe with the help of tools of the Factor distribution.

I hope that makes sense.  I think that is a really important issue.  A number of languages have made decisions which make packaging applications and libraries a serious pain from the perspective of deployment/OS-distribution.

**Comment by mrjbq7** [comment]
Feel free to make a PR that puts this in ``extra/`` and lets people start playing around with the idea!

I'm not ready to make ``curl`` a dependency of basis, but sometime in the future maybe.

**Comment by mrjbq7** [comment]
It's definitely true that we have some challenges if you hope to allow different versions of a single package to co-exist but be named similarly, we would need to re-architect the lookup code to be more dependency-aware. 

For now, having a system that downloads, caches, and adds a vocab root, seems a cool idea to experiment with.

**Comment by lolbinarycat** [comment]
Yeah, this is going to be fairly invasive to fix. *way* to much stuff is generic here, and for no great reason. Like, what even is a "vocab link"? What's with the `platforms.txt` stuff? All the words seem to do multiple things, and it's all a mess.

**Comment by mrjbq7** [comment]
So, your initial suggestion of having a remote vocab-root is not a bad way to start. 

The way vocab paths work is to look through all the roots for a vocab named as requested.  That has the side effect of preventing shadowing.  So you can't have a user vocab-root with a ``math`` vocabulary, for example.

A more generalized approach might be to "root" those vocabularies so ``core:math`` is different from ``user:math``, but that has a lot of potential complexity and verbosity when importing.

Right now if ``REMOTE: https://github.com/mrjbq7/re-factor.git`` had the side effect of downloading and adding a ``re-factor`` root, that would be cool.  Potentially, we could just add that to your ``.vocab-roots`` file.  If you're hoping to do it at "import" time, something like:

```
FROM: https://github.com/mrjbq7/re-factor.git USE: mail-ui
```

Or

```
ROOT: https://github.com/mrjbq7/re-factor.git
USE: re-factor:mail-ui
```

Or something else, that's probably fairly easily possible with small code changes.

Having multiple versions co-exist is probably not easily possible, without having a better dependency tree and keeping multiple versioned vocabularies co-existing...

**Comment by mrjbq7** [comment]
You might want to look at #641 and the gist it links to!

**Comment by lolbinarycat** [comment]
I have read and commented on said gist, don't know if you get notifications for that.

**Comment by AlexIljin** [comment]
Can we please **not** have multiple vocab versions coexisting?
I've got enough of that crap in WinSxS folder, which is a whole new level of DLL hell.

**Comment by odiferousmint** [comment]
Related comment: https://github.com/factor/factor/issues/2350#issuecomment-716085559
